examples/gno.land/r/demo/gnofundme/gnofundme.gno

```diff
@@ -0,0 +1,253 @@
+
+	// the key will be the size of the tree
+	key := campaigns.Size()
+	campaigns.Set(strconv.Itoa(key), campaign)
+
+	return key, "campaign created successfully"
+}
+
+func AddContributor(campaignId string, amount uint, display bool) string {
+	// get the campaign if it exists
+	campaign, ok := campaigns.Get(campaignId)
+	if !ok {
+		return "campaign not found"
+	}
+
+	// get the user address
+	userAddress := std.GetOrigCaller()
+
+	// get the user if he exists
+	user, ok := users.Get(userAddress.String())
+	if !ok {
+		panic("user is not registered")
+	}
+
+	sentCoins := std.GetOrigSend()
+	// amount := sentCoins.AmountOf("ugnot")
+
+	// add the contributor to the campaign
+	campaign.(*pkg.Campaign).AddContributor(user.(*pkg.User), amount, display)
+
+	return "contributor added successfully"
+}
+
+func GetCampaigns() string {
+	if campaigns.Size() == 0 {
+		return "[]"
+	}
+	campaignsArray := make([]*pkg.Campaign, 0, campaigns.Size())
+	campaigns.Iterate("", "", func(key string, value interface{}) bool {
+		c := value.(*pkg.Campaign)
+		campaignsArray = append(campaignsArray, c)
+		return false
+	})
+	return pkg.CampaignsToJSONString(campaignsArray)
+}
+
+func Render(path string) string {
+	if path == "" {
+		return renderHomePage()
+	}
+
+	// split the path
+	parts := strings.Split(path, "/")
+	switch {
+	case len(parts) == 2 && parts[0] == "campaign":
+		return renderCampaign(parts[1])
+	}
+	return "unknown page"
+}
+
+func renderHomePage() string {
+	banker := std.GetBanker(std.BankerTypeRealmIssue)
+	banker.IssueCoin(std.CurrentRealm().Addr(), "ugnot", 350000000)
+	balance := banker.GetCoins(std.CurrentRealm().Addr())
+	// define empty buffer
+	var b bytes.Buffer
+
+	// users counter
+	userc := 1
+	b.WriteString("# Welcome to GnoFundMe\n\n")
+	b.WriteString(ufmt.Sprintf("Balance: %s.\n", balance.String()))
+
+	b.WriteString("# Campaigns :\n\n")
+	campaigns.Iterate("", "", func(key string, value interface{}) bool {
+		c := value.(*pkg.Campaign)
+		campaignURL := ufmt.Sprintf("%scampaign/%s", campaignPrefix, key)
+		b.WriteString(
+			ufmt.Sprintf("## [Campaign %s](%s)\n", key, campaignURL),
+		)
+		b.WriteString(
+			ufmt.Sprintf(
+				"### Description: %s\n", c.GetDescription(),
+			),
+		)
+
+		return false
+	})
+	b.WriteString("# Users:\n")
+
+	users.Iterate("", "", func(key string, value interface{}) bool {
+		u := value.(*pkg.User)
+		b.WriteString(
+			ufmt.Sprintf(
+				"## User #%d: %s\n", userc, u.GetName(),
+			),
+		)
+		b.WriteString(
+			ufmt.Sprintf(
+				"- email: %s\n - address: %s\n", u.GetEmail(), u.GetAddress(),
+			),
+		)
+		userc++
+		return false
+	})
+	return b.String()
+}
+
+func renderCampaign(id string) string {
+	// get the campaign if it exists
+	campaign, ok := campaigns.Get(id)
+	if !ok {
+		return "campaign not found"
+	}
+
+	// define empty buffer
+	var b bytes.Buffer
+	c := campaign.(*pkg.Campaign)
+	b.WriteString(
+		ufmt.Sprintf(
```

Comment:
Maybe changing this ufmt.Sprintf into 1 line instead of 3? To avoid a long path of 42 lines doing the same and reduce it into 18 (less than the half)

Reply:
If you mean the new lines after "ufmt.Sprintf(", I understand. It's because I copied a code as it is from another example. But I'll make the modifications.

Reply:
solved here 5093d30f